Graph based behavior execution scheduling and post processing support

wizgrav opened the pull request:
Following the discussions with @ngokevin @dmarcos @donmccurdy and @fernandojsg :

Enable post processing by extending components with a .tock() method which runs after the scene renders.

A-scene maintains a .renderTarget and switches rendering to that in the presence of at least one tock behavior.

A graph based mechanism is employed to allow for explicitly definable execution ordering for both tick and tock behaviors.

Components can define two way connections in the graph using the properties .precedents and .following, similar to .dependencies.

Sorting to define the execution order takes place only in registerComponent() and batching is done efficiently in a-scene using caches.

The graph based mechanism works fast and comfortable for the developer. Cyclic dependencies in the graph are errors and throw exceptions.

An example utilizing most of the functionality added by this PR at http://wizgrav.github.io/aframe/examples/test/post-processing/index.html

donmccurdy commented:
This is really neat! Nice example.

As far as the API, it feels strange having DOM events hooked into the render loop, firing every 8-16ms. I'm not sure exactly what to recommend though. Other ideas:

```javascript
// 1.
sceneEl.setPostProcessingCallback(function () { /* ... */ });

// 2.
sceneEl.addPostProcessingCallback(function () { /* ... */ });

// 3.
AFRAME.registerComponent('my-component', {
  prerender: function () { /* called only if my-component is attached to sceneEl */ },
  postrender: function () { /* called only if my-component is attached to sceneEl */ },
});
```

wizgrav commented:
@donmccurdy you proposals #1/#2 where also my original thoughts. I went for emit because it seemed to be more consistent with the patterns used in the library. I'm almost certain that performance shouldn't degrade with dom events. Of course this needs investigation but 8/16ms is still a good chunk

wizgrav commented:
Hmm on the other hand I'm not sure if the events are dispatched synchronously

dmarcos commented:
Thanks for tackling this. I don't think emitting events from the render loop is a very good idea. It adds overhead (and most of the time you won't even use it) and compromises performance. Having a way to do postprocessing effects would be indeed very cool. I like more @donmccurdy approach of providing two methods at the component level. You can currently insert your logic in the render loop via the `tick` function of components and systems. It would be nice to have a way to throw warnings or stats on console if  any those user defined functions in the render loop are too slow.

ngokevin commented on src/core/system.js:
```diff
@@ -33,6 +33,16 @@ System.prototype = {
+   * @param {number} time - Scene tick time.
+   * @param {number} timeDelta - Time it took to render the scene.
+   */
+  tick: undefined,
```
We probably shouldn't be using the `system` API to handle post-processing. Systems are sort of global managers for components. We should just use the plain component API.

ngokevin replied:
I don't think we need Systems to have a hook into post-render callback yet.

ngokevin commented on src/core/component.js:
```diff
@@ -66,7 +66,17 @@ Component.prototype = {
+   * @param {number} time - Scene tick time.
+   * @param {number} timeDelta - Time it took to render the scene.
+   */
+  post: undefined,
```
Let's call it `postRender` to be more specific.

ngokevin commented:
Can you remove the `dist` bumps?

ngokevin commented on examples/boilerplate/360-video-post/index.html:
```diff
@@ -0,0 +1,117 @@
+<!DOCTYPE html>
```
Can you move this example to `examples/test/post-processing` and rename the titles/metatags to match?

ngokevin commented on examples/boilerplate/360-video-post/index.html:
```diff
@@ -0,0 +1,117 @@
+    <meta name="description" content="360 Video — A-Frame">
+    <script src="../../../dist/aframe.js"></script>
+    <script id="vertexShader" type="x-shader/x-vertex">
+
```
A lot of whitespace. Can one-line this one:

```html
<script>
  void main() { gl_Position ...
</script>
```

ngokevin commented on examples/boilerplate/360-video-post/index.html:
```diff
@@ -0,0 +1,117 @@
+    </a-scene>
+    
+    <script>
+        AFRAME.registerComponent('effect', {
```
Can you put this in the `<head>` so that it is available before the scene runs?

ngokevin commented:
Hmm, after going over the PR, I'm not sure about piggybacking off the component API. It assumes `postRender` might have a wider use case than just post-processing. Maybe we need a tighter API for register post-processing effects. Thinking about it...

wizgrav commented:
based on @dmarcos comments I added a post method on components. Behaviors are split in tick and post types and handled separately. Performance has indeed improved noticeably(from 50% -> 40% on my box for the example) so this seems like the optimal path to take.

I also added a postPriority property on components. This is intended to facilitate common post proc setups where several components will produce their own renderTargets containing bloom, godrays, ssao etc and then a composer component can pick these up and blend the to screen. Prioritizations happens only on post behaviors(which should only be a handful) and sorting happens only once per frame if there was a change in the post pipeline.

Finally an event is emitted when the sorting happens to indicate that the pipeline has changed. This can be picked up by the composer components so they can reconfigure themselves. I'm not sure about this one maybe a simple timestamp could work. What are your thoughts on this, and the whole pr in general?

wizgrav commented:
@ngokevin sorry for the mess with dist and such, I was having great pain with npm and setting up the build in general. I'll fix everything

wizgrav commented:
I also included the tweaked VREffect inside lib/three.js until it is fixed in three.js itself. I see no reason why they shouldn't do the tweaks themselves. The VREffect render() should match the signature of the base WebGLRenderer since it's a wrapper of it. Another note of interest is the use of depth textures. I think it recently landed on three and fortunately they implemented it as a property on the renderTarget so when aframe upgrades to the latest three.js all that will be needed to utilize depth textures will be

`this.renderTarget.depthTexture = new THREE.DepthTexture()`

and of course resizing it along with the renderTarget itself

ngokevin commented:
I don't like using the component API too much. We are adding a special handler and property that from what I can tell would only be used for post-processing, and then we have to touch play/pause code, and such.

Brainstorming, but here's an alternative API, which would be harder to implement, but would be a bit more explicit. Using the System API, this would be a bit similar to Don's Controls API:

```
AFRAME.registerPostProcessor('some-effect', {
  schema: {},
  vertexShader: 'vertexshaderstring',
  fragmentShader: 'fragmentshaderstring'
});

<a-scene postprocess="bloom, hdr, ssao" bloom="intensity: 0.5>
```

wizgrav commented:
@ngokevin I'm not sure a postproc effect can be described like a shader, it's more of a setup. For instance bloom commonly uses a separable gaussian blur so it would need at least two passes. Good bloom would also entail downsampling several times, blurring, and blending again on the way up. I think components make for a natural expression even if meant to be attached on the scene only. But I don't know I haven't yetlooked over all of aframe's source and patterns so you'd know best

wizgrav commented:
@ngokevin another thing that should be taken into account while brainstorming is that customisation logic ends up in the composer usually. Postproc steps that provide intermediate renderTargets like bloom and ssao tend to be straightforward and so could be cast in place as reusable components. Only the composer needs special treatment, both for doing the composition of all the rest and also to apply the simple effects like vignette aberations, ascii etc At least for production one shouldn't use extra passes for these

ngokevin commented:
Ah, right, thanks for the details. I should clarify the sample API I had would still make post-processing effects a component, but just a subset API of a component. Just because `postRender` + `postRenderPriority` would only be used by effects, seems to make sense to fence it off.

wizgrav commented:
Yeah I understand your concerns. Defining post specific components seems to be the cleaner route then unless there would be a reason for post components to also run some logic at tick time. To be honest though I cannot think of such a case right now, I need to think about it

wizgrav commented:
Ok so I tried to think of cases where having a component tap into both tick and postRender and I have indeed came up with something. I've updated the example which you can see on http://wizgrav.github.io/aframe/examples/test/post-processing/

The alpha channel of the render target can be used to pass information to the post processing chain, the most common case being masking. In the example the pulse component sets up properties on its underlying mesh. The postproc component then traverses the scene and adjusts the opacity according to these properties. It also handles cleanup.

Alternatively the pulse component itself could setup it's opacity by itself but it would need to know the state of post processing. It could do that by listening for something like the "post-modified" event.

I'm not arguing in favor of either approach, it would be convenient to have both options availble.

I would also like to mention something, and I know you're going to hate me @ngokevin , but it really needs to be said. It is inevitable that components will need to be extended. At some point, a component will need to access the results of computations performed on another component on another entity. A way to guarantee the order of execution will be needed. I'm sorry if I've missed something, I still study the source but I think there isn't something like that in currently.

Doing something like postRenderPriority may work for postproc because it's by nature limited in scale but for tick behaviors I don't think it's appropriate at all. The solution unity came up for this situation was to add a LateUpdate method which runs after all update methods. So aframe will also need a tock() to go along with tick(). Again excuse me if I'm missing something. This does not concern this PR, it's something else all together but it's relevant and I feel that it should be taken into account

wizgrav commented:
Following the line of thought from the previous comment, if a tick()/tock() pair is indeed desirable, the api could get some symmetry with a postTick()/postTock() pair for the post processing, getting postRenderPriority out of the picture. Two lanes for the post processing should cover most cases. The third fxaa step I mentioned could be integrated in the composer itself in postTock(). So that setup could also work

wizgrav commented:
@ngokevin @dmarcos Added tock() on components as a straight equivalent of unity's LateUpdate. An equivalent pair is implemented for post render as postRenderTick, postRenderTock. These should be enough to guarantee the order of execution so postRenderPriority was removed. The post-modified event is still emitted as it is useful for reconfiguring the compositor when the post chain changes.

The example at http://wizgrav.github.io/aframe/examples/test/post-processing/ was commented and updated to reflect all functionality in the PR. Some crude bloom was added to highlight the tick/tock pattern, you can toggle it by right clicking.

ngokevin commented:
Thanks, will take a look when I can

ngokevin commented:
Can you remove the `dist` files from the PR?

ngokevin commented on examples/test/post-processing/index.html:
```diff
@@ -0,0 +1,309 @@
+    <script src="../../../dist/aframe.js"></script>
+    
+    <script>
+        AFRAME.registerComponent('pulse', {
```
Let's register these components in separate files: e.g., `components/pulse-effect.js`, `components/envmap.js`, `components/bloom.js`

wizgrav commented:
Hmm, I split the example but I'm not sure I did the right thing with the dist files. I did a checkout on the last commit before I touched them. Can you suggest how could they get out of the PR completely?

There's also the issue with VREffect being included in lib/three.js which is not so good. I can try making a PR to three so they can fix it on their source but to be honest, requiring such a core part of aframe from an examples/ folder kind of makes me nervous. Since you control the revision of three included anyway, maybe you should consider including VREffect.js in aframe's source somewhere.

ngokevin commented on examples/test/post-processing/components/compositor.js:
```diff
@@ -0,0 +1,141 @@
+  ].join("\n"),
+    
+  fragmentText: [
+    "uniform sampler2D srcTexture;",
```
nice, thanks

ngokevin commented:
- `dist/` files are out and look OK now.
- Yeah, that's not great. But @dmarcos just pulled them into `vendor/` for WebVR 1.0 API reasons. We still wouldn't want to touch them. Maybe we need that module to live in its own home?

ngokevin commented:
Needs a rebase. I'm not sure how we should handle the `VREffect` change collisions. Ideally, all the changes would be upstreamed to three.js, but maybe we should just move it to `lib/` and maintain it for now...

wizgrav commented:
Yes I agree. Do you want me to do it here?

ngokevin commented on examples/test/post-processing/index.html:
```diff
@@ -0,0 +1,36 @@
+    </a-scene>
+      
+    <script>
+        var scene = document.querySelector("a-scene");
```
This can be a scene component like `components/bloom-toggler.js`

wizgrav replied:
Not sure I got that, you mean renaming the file and component or something else?

ngokevin replied:
I mean move the whole `script` logic to a component

wizgrav replied:
Ok got it you mean the last script on bottom, yeap will fix that

ngokevin commented on examples/test/post-processing/index.html:
```diff
@@ -0,0 +1,36 @@
+<html>
+  <head>
+    <meta charset="utf-8">
+    <title>Post Process</title>
```
"Post-processing" here and below would match better

ngokevin commented on src/core/a-entity.js:
```diff
@@ -683,7 +683,9 @@ function isComponentMixedIn (name, mixinEls) {
   component.pause();
   // Remove tick behavior.
-  if (!component.tick) { return; }
+  if (!component.tick && !component.tock && !component.postRenderTick && !component.postRenderTock) {
```
Hm, maybe a separate function

```js
function hasBehavior (component) {
  return component.tick...;
}
```

wizgrav replied:
Yeap this line has gotten way too long

ngokevin commented on src/core/scene/a-scene.js:
```diff
@@ -118,8 +124,15 @@ var AScene = module.exports = registerElement('a-scene', {
         var behaviors = this.behaviors;
-        if (behaviors.indexOf(behavior) !== -1) { return; }
-        behaviors.push(behavior);
+        for (var b in behaviors) {
```
We usually use `Object.keys().forEach` pattern for object enumeration, which I read to be a bit faster.

ngokevin replied:
avoid 1-letter var names: s/b/behaviorType

wizgrav replied:
forEach makes me weary as it would create a closure but I'm not really sure how the browsers will optimise it. I went on to make a test case in jsperf but it seems their site is down for maintenance. I'll try again later

ngokevin replied:
I've read `for...in``` makes hasOwnProperty calls for each key. But that's just from skimming SO

wizgrav replied:
The body of the function is so short that I think in either case it will be optimised to the same native code. I'll switch to forEach for consistency

ngokevin commented on src/core/scene/a-scene.js:
```diff
@@ -118,8 +124,15 @@ var AScene = module.exports = registerElement('a-scene', {
-        if (behaviors.indexOf(behavior) !== -1) { return; }
-        behaviors.push(behavior);
+        for (var b in behaviors) {
+          var bs = behaviors[b];
```
s/bs/behaviorTypeArr or behaviorArr or something

ngokevin commented on src/core/system.js:
```diff
@@ -34,6 +34,16 @@ System.prototype = {
    */
   tick: undefined,
 
+ /**
```
Do we need tock support for systems yet?

ngokevin commented on src/core/scene/a-scene.js:
```diff
@@ -57,7 +57,13 @@ var AScene = module.exports = registerElement('a-scene', {
+        this.behaviors = {
+          tick: [],
+          tock: [],
+          postRenderTick: [],
```
Maybe `tick`, `tock`, `tickPostRender`, `tockPostRender`?

wizgrav replied:
I don't mind with either option. Your call

ngokevin replied:
Probably go with the option I had to namespace it

ngokevin commented on src/core/scene/a-scene.js:
```diff
@@ -317,10 +395,33 @@ var AScene = module.exports = registerElement('a-scene', {
-        this.renderer.render(this.object3D, camera);
 
         this.time = time;
+
```
extra line

ngokevin commented on src/core/scene/a-scene.js:
```diff
@@ -317,10 +395,33 @@ var AScene = module.exports = registerElement('a-scene', {
+          this.tock(time, timeDelta);
+        }
+
+        if (this.behaviors.postRenderTick.length > 0 || this.behaviors.postRenderTock.length > 0) {
```
need the `> 0`?

wizgrav replied:
not really, should I take it out?

ngokevin replied:
yeah

ngokevin commented on src/core/scene/a-scene.js:
```diff
@@ -317,10 +395,33 @@ var AScene = module.exports = registerElement('a-scene', {
+
+          this.renderer.render(this.object3D, camera, this.renderTarget);
+
+          this.time = time;
```
Can put all the `time` line next to each other:

```
this.time...
time = ...
timeDelta = ...
```

wizgrav replied:
On a related note @dmarcos mentioned that he'd wanted a means to check if post processing takes too long. Should we take another timestamp with performance.now() and I don't know spew some debug with that here  or should that be done inside a component in postRenderTock?

ngokevin replied:
As long as the performance of non-post-processed scenes don't degrade, it's OK.

wizgrav replied:
it won't even run if there's not post processing but what kind of debug should we output? I don't like the idea of outputing on the console every frame. Three.js has a stats object inside where it collects various metrics like draw calls etc Maybe we should do something similar?

ngokevin replied:
Try out `<a-scene stats>` and be amazed

wizgrav replied:
awesome I totally missed that, so should I just window.performance.mark it?

ngokevin commented:
Bunch of `npm run lint` errors

wizgrav commented:
the bulk of lint errors come from lib/three.js cause of the inclusion of VREffect.js, I'll take it on it's own file out as we said but should we ignore it in lint checks?

ngokevin commented:
Yeah, still need to resolve it with the VREffect in `vendor/`, but there'll be a lot of lint errors for `a-scene`, and all the example JS files.

wizgrav commented:
@ngokevin  Rebased and squashed. Did also the minor changes needed for renderTarget support in @dmarcos  VREffect.js in vendor/

ngokevin commented on vendor/VREffect.js:
```diff
@@ -232,7 +232,7 @@ THREE.VREffect = function ( renderer, onError ) {
 	cameraR.layers.enable( 2 );
 
-	this.render = function ( scene, camera ) {
+	this.render = function ( scene, camera, renderTarget, forceClear ) {
```
Can we upstream this to three.js?

wizgrav replied:
I believe so, this is the signature of WebGLRenderer.render() and since VREffect behaves similarly it should match it.

ngokevin replied:
Want to do it or me? https://github.com/mrdoob/three.js/blob/master/examples/js/effects/VREffect.js

ngokevin commented on src/lib/three.js:
```diff
@@ -26,3 +26,4 @@ require('../../vendor/VRControls');  // THREE.VRControls
 require('../../vendor/VREffect');  // THREE.VREffect
 
 module.exports = THREE;
+
```
whiteline

ngokevin commented on src/core/scene/a-scene.js:
```diff
@@ -312,10 +397,36 @@ module.exports = registerElement('a-scene', {
-        this.effect.render(this.object3D, camera);
 
         this.time = time;
+
```
whitespace

ngokevin commented:
@fernandojsg r?

ngokevin commented on src/core/scene/a-scene.js:
```diff
@@ -312,10 +397,36 @@ module.exports = registerElement('a-scene', {
+          this.tock(time, timeDelta);
+        }
+
+        if (this.behaviors.tickPostRender.length || this.behaviors.tockPostRender.length) {
```
How about just a `behavior-added` event when you add a behavior?

ngokevin commented on src/core/scene/a-scene.js:
```diff
@@ -312,10 +397,36 @@ module.exports = registerElement('a-scene', {
+
+        if (this.behaviors.tickPostRender.length || this.behaviors.tockPostRender.length) {
+          // Post processing setup has been modified, emit a notification.
+          if (this.postNeedsNotify) {
```
How about just a behavior-added event when you add a behavior?

wizgrav replied:
Sure but it only applies to post behaviors, I think doing it for ticks especially would cause lots of traffic. It's also emmited when a post behavior is removed. I don't think separate events are needed for these. The compositor will mainly listen to these in order to recompile itself when something changes in the post chain. Maybe "post-behavior-changed" or something like that?

ngokevin commented on src/core/scene/a-scene.js:
```diff
@@ -312,10 +397,36 @@ module.exports = registerElement('a-scene', {
+
+          this.time = time;
+          time = window.performance.now();
+          timeDelta = time - this.time;
```
I think we can just reuse the timeDelta from before. Makes sense that it's timeDelta between each frame

wizgrav replied:
Sure, to be honest I can't think of a case for using the second timedelta besides knowing how long the base rendering took, maybe we could just mark for stats purposes?

ngokevin commented on src/core/scene/a-scene.js:
```diff
@@ -299,6 +334,56 @@ module.exports = registerElement('a-scene', {
+    tickPostRender: {
+      value: function (time, timeDelta) {
+        // Components.
+        this.behaviors.tickPostRender.forEach(function (component) {
```
Might want to abstract this to a function outside of the scene prototype below:

```
runBehavior('tickPostRender', time, timeDelta)

// ...

function runBehavior (type, time, timeDelta)
```

wizgrav replied:
Yeah good idea, lots of duplication in there

wizgrav replied:
hmmm, not all behaviors are handled the same. The post behaviors don't care if the component's el isPlaying, I think we said that it shouldn't matter for these? As it is right now even if the scene is not playing the base rendering still happens so post render should as well

ngokevin commented on src/core/scene/a-scene.js:
```diff
@@ -56,7 +56,13 @@ module.exports = registerElement('a-scene', {
+          tickPostRender: [],
+          tockPostRender: []
+        };
+        this.postNeedsNotify = false;
```
Yeah, let's just emit right when we add/remove rather than keeping this state var and emitting during render loop

wizgrav replied:
Well this will lead to several events handled where one would be sufficient, usually you'll pass several components to setup the chain at once. The compositor will catch the events and reconfigure itself. It's not really a problem as this will generally not happen often but If you mean to generalize the notification events for all types of behaviors then this will probably lead to alot of event traffic.

ngokevin replied:
Perhaps just an event for added/removed post-processing behaviors then? Not too worried about batching them.

wizgrav replied:
Ok I'll take out the dirty flag and emit in add/remove behavior, how does "post-behaviors-changed" sounds?

ngokevin replied:
Yeah or `modified`

fernandojsg commented:
I'm looking around the API and I'm thinking how to deal with multiple postprocessing on the scene.
Sometimes you may want to add gloom + ao for the whole scene but also want to active from time to time other effects like nightvision or some kind of underwater distorsion.
Ideally you could be able to chain them, switch on/off and change the parameters for each one.
Did you already think about a way to solve this with the current implementation?
I was thinking on something like:
```
<a-postprocessing>
    <a-effect id="sparkles1" type="sparkles" img="sparkes1.jpg" opacity="0.5" color="#fff">
    <a-effect id="sparkles2" type="sparkles" img="sparkes2.jpg" opacity="1.0" color="#f00">
    <a-effect id="gloom" type="gloom" intensity="1.2" radius="2.0">
    <a-effect id="nightvision" type="nightvision" intensity="2.0" noise="1.2">
</a-postprocessing>
```
Everyone will be loaded but just some of them will be used, defined by the `post-process` param on `a-scene`, and just the last one is the one that will render back to the framebuffer.
```
<a-scene post-process="gloom, nightvision"/>
```
So you could add or remove effects from the postprocessing chain modifying `post-process` effect. And also you could modify the parameters for each effect accessing directly to the element defining it.

ngokevin commented:
Those individual effects could be components rather than adding a new element too.

@wizgrav I'm rethinking the tick/tock...It would be more flexible to explicitly define an order like you had before right?

fernandojsg commented:
but if you want to reuse some like the `sparkles` but with different parameters, you need to create a new component? or will you create a `a-mixin` for each one?

dmarcos commented on examples/test/post-processing/components/bloom.js:
```diff
@@ -0,0 +1,111 @@
+https://software.intel.com/en-us/blogs/2014/07/15/an-investigation-of-fast-real-time-gpu-based-image-blur-algorithms
+*/
+
+AFRAME.registerComponent('bloom', {
```
Can we reuse the registerShader API to register post processing effects?

wizgrav replied:
We could but it would be limiting. A postproc effect usually employs several shaders and passes to produce. For instance an ssao pass first runs the ssao shader which gets out noisy so it also applies a special blur to it, taking depth into account so as to avoid spilling the ao to near objects. The components are the best solution imo, and with the graph mechanism could be used in creative ways, like defining entities that each contain a whole chain of postprocessing filters and combine these entities

dmarcos commented on examples/test/post-processing/index.html:
```diff
@@ -0,0 +1,27 @@
+    <script src="components/bloom-toggler.js"></script>
+  </head>
+  <body>
+    <a-scene post-process="" bloom-toggler="">
```
If we reuse the register shader API to register effects one could do 
````html
<a-scene post-process="effects: bloom, nightVision; bloomIntensity: 1.0; nightVisionIntensity: 0.6">...</a-scene>
````
The name of the effect is prepended to the attribute names defined in the schema of the effect (aka shader)

dmarcos commented:
What about this:
````html
<a-scene post-process="effects: sparkles, sparkles, bloom, nightVision; bloomIntensity: 1.0; nightVisionIntensity: 0.6; sparkles1Color: red; sparkles2Color: blue">...</a-scene>
````
The name of the effect and the index are automatically prepended based on the schema of the effect and the order in the list.

ngokevin commented:
I think we should decide on a post-processing API later. This PR is really about being able to order component `tick` calls and being able to hook into the post-render. That is very flexible for now, and when we have more experience, we can maybe define an explicit PP API later.

dmarcos commented:
I think the post-processing component should be the one calling the tick methods of each effect.

ngokevin commented:
That could be layered in or decided later. Note that PP is just a component in `examples/`; it's just one use case. The PR is actually about ordering components. Another use case for that is the controls API needs support for ordering `controls` before `velocity`.

dmarcos commented:
What I mean is I think we just need a new method tock called after the render call and that's all. If there are multiple tick functions to call on a specific order should be managed by the component post-processing. We don't need a generic API at this point. What other use case are there? The API proposed is over killing in my opinion.

ngokevin commented:
So if we had two components `controls` and `velocity`, where `controls` needs to be run before `velocity`, we wouldn't be able to control that with just `tick`/`tock`. We need a way to order those `tick`s as well. That could be a different PR though, yeah.

Yeah, I think the bare minimum for this PR, is a `tock` and events on `add/removeBehavior`. I think PP could be accomplished within those constraints?

dmarcos commented:
You can express that with the dependencies components. If a component is a dependency it will be initialized in first place and the tick method called before the dependent component one.

fernandojsg commented:
As the active PP `effects` could be dynamically modified I believe the order must be controlled from a `post-process` component that is the responsible to access the PP system. I'm not sure why do you want to implement `tick/tock` for PP instead of doing a composer similar to the one used in threejs' `THREE.EffectComposer` or Jaume's `Wagner`. So you could just inject a call in the `a-scene` render when postprocessing is active and the order, passes and so on will be delegated to the `post-process` components, instead of worry about `tick` order for each effect's component.

wizgrav commented:
@fernandojsg This is how I think the post processing could work with both flexibility and efficiency. Here's how the markup would look like

`<a-scene post-process="gloom, nightvision" gloom="intensity:0.5" nightvision="intensity:0.33"></a-scene>`

In this imagine that gloom would generate a texture that needs to be additively blended with the base rendering but nightvision is a simple equation applied to the frag color. The gloom component would do it's processing in tickPostRender and produce the texture. It would also define a postInfo property(not part of the components prototype) that would contain a unique id, uniform definitions for intensity etc and the fragment shader text it needs for the composition later kind of like

`postInfo: {
  name: "gloom",
  uniforms: {
    "srcTexture": {type: "t", value: this.gloomTexture} // Produced by the component
    "intensity": {type: "1f", value: 0.5} //Settable by the component in it's update()  
  },
  fragmentText: " void gloomPostProcess(in out vec4 currentColor, vec2 uv){ currentColor += texture2D(gloomInfo.srcTexture, uv) * gloomInfo.intensity; }"
}`

The "post-process" component, the compositor, would collect this information and create on shader with a setup like

`{
  uniforms: {
    "gloomInfo.srcTexture": gloomComponent.uniforms.srcTexture,
    "gloomInfo.intensity": gloomComponent.uniforms.intensity
  },
  fragmentShader: "
   sample2D sceneTexture; // scene.renderTarget
   varying vec2 vUv;
    struct gloomInfoStruct
    {
       sampler2D srcTexture;
       float intensity;
    } gloomInfo; // This is generated by the compositor based on the uniforms from gloom component

    void gloomPostProcess(in out vec4 currentColor, vec2 uv){ currentColor += texture2D(gloomInfo.srcTexture, uv) * gloomInfo.intensity; } // This comes from the gloom component itself

    void main(){
        vec4 color = texture2D(sceneTexture, vUv);
        gloomPostProcess(color,vUv);
        nightvisionPostProcess(color, vUv);
        gl_FragColor = vec4(color.rgb,1.0);
    }
  "
}`

Nightvision would not need to produce a texture so it would just provide uniforms for control and it's shader chunk. When you change attributes text components would set their uniforms in their update() and the compositor who has them by reference would apply them with no further intervention

fernandojsg commented:
@wizgrav With that approach you're avoiding more drawcalls but if the effects change dynamically (going inside/outside water, nightvision on/off) you're rebuilding the whole shader on each switch which is not very performant :\
Also the `fragmentText` should be defined in that way by the user? or it's dynamically generated from the original fragment shader from the user?

wizgrav commented:
when you say switch, you mean taking out the nightvision completely or just playing with it's intensity to make it apply more or less? In the former case yes the post chain will change and the compositor shader will need to be recompiled but if you just want to toggle the effect setting the intensity will just update the uniform. 

The fragmentText will be part of the gloomComponent's definition. The compositor will generate the gloomInfo struct text and include the fragmentText from gloomComponent + dynamically generate it's main as a chain. Some conventions are used, the components will have to include a unique prefix (postInfo.name) to aid the compositor for it's dynamic generation of text. 

The fragmentText of gloomComponent will be defined with these conventions in mind, that it will have access to a gloomInfo struct based on it's unique name containing it's uniforms and it will need to provide all the rest and name it's main function as gloomPostProcess so the compositor can call it

fernandojsg commented:
I believe it will be quite hard to come out with an hybrid system to control the passes in order to optimize the drawcalls as creating just one whole shader is not possible for many effects so it should be necessary to create a logic to separate the "full" passes from the "composed" passes as you described.
For example if you want to apply blur, it's way much faster to do two passes with horizontal blur -> rotate -> horizontal blur, instead of one pass. Also you will have some passes that will be applied to RT of different sizes, so the system should be intelligent enough to deal with this...
Maybe you could define which passes should be executes separately in the API to prevent overengineering it.

wizgrav commented:
Yes you can check out exactly that with bloom in the example. It's actually doing it in 5 passes(I wanted to try out kawase bloom) Components that need to create textures will do that in their tickPostRender(). They'd still need to provide some fragmentText for the blending. Bloom would need to add on top while ssao for instance would need to modulate. The compositor would be the only component that defines postRenderTock so it would run after all the others. 

There's still one caveat. Some effects need the final composited frame to work on. Fxaa or glitch for instance. For these instead of a postInfo, the component will provide a function with a signature of postProcess(inRenderTarget, outRenderTarget) and again it will be the compositor who will apply them after he's done with the uber shader

wizgrav commented:
We could have the mechanism I described as standard for post processing in frame. I like the idea that it could instead be implemented on top of the existing component infrastructure though. This would make it easily exchangeable if a better scheme comes up afterwards.

wizgrav commented:
@dmarcos @ngokevin is right, the postproc related stuff, addition of a .renderTarget and a switch to render on that instead of the screen are pretty much imposed by the apis provided by three itself. This PR has come to mostly concern with defining priorities for the execution of behaviors that need to run every frame. I'm leaning towards the, not so flexible, tick/tock mechanism because of the precedent in unity. It seems like a safe choice. Of course just because unity did it doesn't mean it is the final answer

wizgrav commented:
@dmarcos reusing the dependencies property you mentioned makes sense and is attractive as an option. It is not enough as it is though as it ensures that controls will run before position for the components attached on a single entity only where what we really need is all control behaviors from all components executing before the first position component of any entity. Maybe not for the specific case but you should check unity's LateUpdate() for the rationale behind the need for ordering behavior groups/types. It's all about entity data interchange. 

We could accomplish that by making scene.behaviors a dynamically updating structure adding new symbols as they appear in addBehavior(), sorting only when a new symbol gets registered in runtime so as to be efficient. There may be some issues with circular dependencies but those will be error cases anyway. 

For post processing we could use tock() to tap on the post render side, allow systems to tap into it as well and define the compositor as a system to ensure it would run after all post component behaviors. I'd prefer this scheme than having to define numeric priorities, what do you think @ngokevin?

wizgrav commented:
Following the discussions with @ngokevin @dmarcos @donmccurdy and @fernandojsg :

Enable post processing by extending components with a .tock() method which runs after the scene renders.

A-scene maintains a .renderTarget and switches rendering to that in the presence of at least one tock behavior.

A graph based mechanism is employed to allow for an explicitly definable execution order for both tick and tock behaviors.

Components can define two way connections in the graph using the properties .precedents and .following, similar to .dependencies.

Sorting to define the execution order takes place only in registerComponent() and batching is done efficiently in a-scene using caches.

The graph based mechanism works is fast and comfortable for the developer. Cyclic dependencies in the graph are errors and throw exceptions.

Post processing adapts naturally to the graph and becomes flexible and easy, aided by a-scene's new methods .getBehaviors() and .getPrecedence().

An example utilizing most of the functionality added by this PR can be seen at http://wizgrav.github.io/examples/test/post-processing/index.html

wizgrav commented:
While fixing the tests, I found some issues that you should take note of @ngokevin @dmarcos. In a-entity.test.js components are manually cleared with "components[componentName] = undefined". I thought that this was the same as "delete components[componentName]". Apparently it isn't, the keys are still enumerable with an undefined value. I iterate the component keys to do the topological sort and undefined values were not handled. Should I handle these or just fix the tests to use delete instead?

ngokevin commented:
Yeah, we can fix the tests to `delete`

ngokevin commented:
The `precedents`/`following` ain't too bad. It's better than just hijacking the `dependencies` API because that is meant for a different purpose (i.e., controls initialization rather than tick execution). It's also better than assigning numbers because it's much tighter and explicit.

Will have to look at the graph-based API because it is fairly complex. But sounds cool.

wizgrav commented:
As the dependencies property was already defining a graph, the concept is not really foreign to the framework, it's more of an upgrade and generalization of an existing mechanism

wizgrav commented:
Tidied it up, graphing related functionality is isolated in utils/precedence.js. The most drastic changes are in a-scene, the batching of components, but at least they have been done with potential future generalization in mind. The way the graph sorting is setup already makes it generally applicable. It can be used with systems and could even be used for shaders, geometries and anything that maps names to definitions. Wherever it makes sense it would be trivial to add it.

ngokevin commented:
Thanks. We'll try to get to this and understand it. Though it's Memorial Day Weekend, I'll be on a trip next week, and we'll all be in London for a week right after. Focuses will be on controls, documentation, and 0.3.0 release, but I hope we can get to this!

wizgrav commented:
Graphing is generally useful so I took it out into it's own package https://www.npmjs.com/package/precedence-maps and pull it as a dependency. The package also provides ordered stores so the code in a-scene is much cleaner now

ngokevin commented on tests/core/scene/a-scene.test.js:
```diff
@@ -110,7 +110,7 @@ helpers.getSkipCISuite()('a-scene (with renderer)', function () {
   test('calls behaviors', function () {
     var scene = this.el;
-    var Component = { el: { isPlaying: true }, tick: function () {} };
+    var Component = {el: { isPlaying: true }, tick: function () {}};
```
you're my type of guy

ngokevin replied:
also around the isPlaying too

ngokevin commented on src/core/system.js:
```diff
@@ -75,6 +85,7 @@ module.exports.registerSystem = function (name, definition) {
   NewSystem.prototype = Object.create(System.prototype, proto);
   NewSystem.prototype.name = name;
   NewSystem.prototype.constructor = NewSystem;
+
```
no need

ngokevin commented on src/core/scene/a-scene.js:
```diff
@@ -313,9 +367,24 @@ module.exports = registerElement('a-scene', {
+        window.performance.mark('render-iteration-finished');
 
         this.time = time;
+
```
no need

ngokevin commented on src/core/scene/a-scene.js:
```diff
@@ -313,9 +367,24 @@ module.exports = registerElement('a-scene', {
         }
-        this.effect.render(this.object3D, camera);
+
+        window.performance.mark('render-iteration-started');
```
can get rid of the whitespace lines until `render-iteration-finished` and move the comment above this entire block such that it's clear this whole block is about PP

ngokevin commented on src/core/scene/a-scene.js:
```diff
@@ -212,6 +236,10 @@ module.exports = registerElement('a-scene', {
           alpha: true
         });
+
+        this.renderTarget = new THREE.WebGLRenderTarget(window.innerWidth, window.innerHeight, { minFilter: THREE.LinearFilter, magFilter: THREE.NearestFilter, format: THREE.RGBAFormat });
```
let's start steering towards no whitespace around `{}`

ngokevin commented on examples/index.html:
```diff
@@ -229,6 +229,7 @@
       <li><a href="test/video/">Video</a></li>
       <li><a href="test/videosphere/">Video 360&deg;</a></li>
       <li><a href="test/visibility/">Visibility</a></li>
+      <li><a href="test/post-processing/">Post Processing</a></li>
```
alphabetize

ngokevin commented:
Do you think it'd be a good idea to split this PR up into three parts so it'd be easier to review?

- `tock`
- execution ordering
- post processing exaple

wizgrav commented:
@ngokevin yes I was also thinking about splitting it. The only issue is that the PRs would conflict with each other. They both touch this.behaviors in a-scene.js which is split in tick and tock. We could do the ordering as the first PR for tick() and then do tock(). 

Should the example for tock be a PR on it's own?

ngokevin commented:
Is it possible to do `tock` addition first? I think that would be easiest. And then later, we can layer on the ordering?

wizgrav commented:
it's certainly possible but mind that the example actually depends on the order guarantees. As the bloom component gets added/removed by bloom-toggler it's execution order will change. This means that the compositor component(post-process) which outputs to screen will blend the current scene's frame with the bloom texture from the previous frame. We can probably workaround that by also removing and readding the post-process in order as well

ngokevin commented:
Yeah, I was suggesting that we save the post-processing example for last.

wizgrav commented:
Ok I'll start off with a new PR for tock() then

ngokevin commented:
Cool. We can keep this one open to track the overall work.

cvan commented:
@wizgrav did you want to wait until PR #1564 is merged before fixing this one up too?

wizgrav commented:
@cvan the tock() method and the behavior ordering overlap on some parts. Should I merge both in this one?

wizgrav commented:
@cvan #1564 is rebased and clear for merge

cvan commented:
@wizgrav thanks. @ngokevin and @dmarcos are best to review #1564. thanks; excited to see this get merged 👍

ngokevin commented:
I think we can close. Not much demand for ordering past tick/tock and it seems aframe-effects was able to get by.
